modules/chemical_reactions/src/actions/ChemicalCompositionAction.C

```diff
@@ -0,0 +1,208 @@
+
+    for (const auto i : index_range(_phases))
+      _problem->addAuxVariable(aux_var_type, _phases[i], params);
+
+    for (const auto i : index_range(_species))
+      _problem->addAuxVariable(aux_var_type, _species[i], params);
+
+    for (const auto i : index_range(_element_potentials))
+      _problem->addAuxVariable(aux_var_type, _element_potentials[i], params);
+  }
+
+  //
+  // Set up initial conditions from a file
+  //
+  if (_current_task == "add_ic" && isParamValid("initial_values"))
+  {
+    readCSV();
+    for (auto it : _initial_conditions)
+    {
+      const std::string class_name = "ConstantIC";
+      auto params = _factory.getValidParams(class_name);
+      params.set<VariableName>("variable") = it.first;
+      params.set<Real>("value") = it.second;
+      _problem->addInitialCondition(class_name, it.first + "_ic", params);
+    }
+  }
+
+  //
+  // Initiate Chemistry Model
+  //
+  if (_current_task == "add_user_object")
+  {
+    // initialize database in Thermochimica
+    if (isParamValid("thermofile"))
+    {
+      const auto thermo_file = getParam<FileName>("thermofile");
+      if (thermo_file.length() > 120)
+        paramError("thermofile",
+                   "Path exceeds thermiochimica's maximal permisible length of 120 with ",
+                   thermo_file.length(),
+                   " characters: ",
+                   thermo_file);
+
+      Thermochimica::setThermoFilename(thermo_file);
+
+      // Read in thermodynamics model, only once
+      Thermochimica::parseThermoFile();
+
+      int idbg = Thermochimica::checkInfoThermo();
+      if (idbg != 0)
+        paramError("thermofile", "Thermochimica data file cannot be parsed. ", idbg);
+
+      Thermochimica::checkTemperature(_tunit);
+      Thermochimica::checkPressure(_punit);
+      Thermochimica::checkMass(_munit);
+
+      Thermochimica::setUnitTemperature(_tunit);
+      Thermochimica::setUnitPressure(_punit);
+      Thermochimica::setUnitMass(_munit);
+    }
+  }
+
+  //
+  // do we need those?! (yes, for now we do)
+  //
+  if (_current_task == "add_aux_kernel")
+  {
+    auto params = _factory.getValidParams("SelfAux");
+    params.set<ExecFlagEnum>("execute_on") = {EXEC_INITIAL, EXEC_TIMESTEP_END};
+
+    for (const auto i : index_range(_phases))
+    {
+      const std::string ker_name = _phases[i];
+      params.set<AuxVariableName>("variable") = ker_name;
+      _problem->addAuxKernel("SelfAux", ker_name, params);
+    }
+
+    for (const auto i : index_range(_species))
+    {
+      const std::string ker_name = _species[i];
+      params.set<AuxVariableName>("variable") = ker_name;
+      _problem->addAuxKernel("SelfAux", ker_name, params);
+    }
+
+    for (const auto i : index_range(_element_potentials))
+    {
+      const std::string ker_name = _element_potentials[i];
+      params.set<AuxVariableName>("variable") = ker_name;
+      _problem->addAuxKernel("SelfAux", ker_name, params);
+    }
+  }
```

Comment:
umm so does this mean the other object does not work standalone?

Reply:
what other object?

Reply:
the auxkernel

Reply:
No it does not. My suspicion is that MOOSE does not see the other variables as being set by a proper kernel and does not put them in the list to be added back to the solution vector. SelfAux is the dummy kernel that triggers the write back.

Reply:
is there a way to prevent a user from building this auxkernel then

Reply:
I don't understand which aux kernel you mean, and why we would want to prevent the user from building it.

Reply:
We would want to prevent the user from building a ThermochimicaAux without the ThermochimicaNodalData because it will not perform as intended (the writing issue)

Reply:
No we would not want the user to use a ThermochimicaAux without SelfAux on the other variables. Imo it is not worth doing that check as I'm planning to fix/change the aux write stuff.

Reply:
What's the time horizon on that fix?

Reply:
It is my unleashed project in the week before Christmas.

Reply:
https://github.com/idaholab/moose/issues/22563